File: combinationsum/combination_sum.go
package combinationsum

import (
	"reflect"
	"sort"
)

var solutions [][]int

func clearSolutions() {
	solutions = make([][]int, 0)
}

// CombinationSum ...
func CombinationSum(candidates []int, target int) [][]int {

	defer clearSolutions()
	var solution []int
	backTrack(candidates, target, solution, 0)

	return solutions
}

func contains(solutions [][]int, solution []int) bool {
	var contains bool

	for _, solutionFromSolutions := range solutions {
		if reflect.DeepEqual(solution, solutionFromSolutions) {
			contains = true
			break
		}
	}

	return contains
}

func backTrack(candidates []int, remains int, solution []int, startIndex int) {

	if remains < 0 {
		return
	} else if remains == 0 {
		solutionToAppend := make([]int, len(solution))
		copy(solutionToAppend, solution)
		sort.Ints(solutionToAppend)
		if !contains(solutions, solutionToAppend) {
			solutions = append(solutions, solutionToAppend)
			return
		}

	} else {
		for ind := range candidates {
			solution = append(solution, candidates[ind])
			backTrack(candidates, remains-candidates[ind], solution, ind)
			solution = solution[:len(solution)-1]
		}
	}
}

File: combinationsum/combination_sum.py
from typing import List, Optional
import logging
import copy

class Solution:

    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:

        def backtrack(candidates: List[int], solution: List[int], start: int, remains: int) -> None:

            if remains < 0:
                return
            if remains == 0:
                solutions.append(copy.deepcopy(solution))
                return
            
            for ind in range(start, len(candidates)):
                solution.append(candidates[ind])
                backtrack(candidates, solution, ind, remains-candidates[ind])
                solution.pop(-1)

        solutions = list()
        if target:
            backtrack(candidates=candidates, solution=[], remains=target, start=0)
        
        return solutions



if __name__ == "__main__":
    logging.getLogger().setLevel(logging.INFO)
    print(Solution().combinationSum(candidates=[2, 3, 6, 7], target=7))
    print(Solution().combinationSum(candidates=[2, 3, 5], target=8))
